dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/EventOperationParamsMapper.java

```diff
@@ -237,19 +247,128 @@ private OrganisationUnit validateOrgUnit(String orgUnitUid) throws BadRequestExc
   }
 
   private void validateUser(
-      User user, Program pr, ProgramStage ps, String orgUnitUid, OrganisationUnit orgUnit)
+      User user,
+      Program program,
+      ProgramStage programStage,
+      OrganisationUnit orgUnit,
+      OrganisationUnitSelectionMode ouMode)
       throws ForbiddenException {
-    if (pr != null && !user.isSuper() && !aclService.canDataRead(user, pr)) {
-      throw new ForbiddenException("User has no access to program: " + pr.getUid());
+    if (user.isSuper()) {
+      return;
+    }
+    if (program != null && !aclService.canDataRead(user, program)) {
+      throw new ForbiddenException("User has no access to program: " + program.getUid());
+    }
+
+    if (programStage != null && !aclService.canDataRead(user, programStage)) {
+      throw new ForbiddenException("User has no access to program stage: " + programStage.getUid());
     }
 
-    if (ps != null && !user.isSuper() && !aclService.canDataRead(user, ps)) {
-      throw new ForbiddenException("User has no access to program stage: " + ps.getUid());
+    if (orgUnit != null
+        && getUserAccessibleOrgUnits(
+                program,
+                user,
+                orgUnit,
+                ouMode,
+                organisationUnitService::getOrganisationUnitWithChildren)
+            .isEmpty()) {
+      throw new ForbiddenException("User does not have access to orgUnit: " + orgUnit.getUid());
     }
+  }
+
+  /**
+   * Returns a list of all the org units the user has access to
+   *
+   * @param program the program the user wants to access to
+   * @param user the user to check the access of
+   * @param orgUnit parent org unit to get descendants/children of
+   * @param orgUnitDescendants function to retrieve org units, in case ou mode is descendants
+   * @return a list containing the user accessible organisation units
+   */
+  private List<OrganisationUnit> getUserAccessibleOrgUnits(
+      Program program,
+      User user,
+      OrganisationUnit orgUnit,
+      OrganisationUnitSelectionMode orgUnitMode,
+      Function<String, List<OrganisationUnit>> orgUnitDescendants) {
+
+    if (orgUnitMode == null) {
+      if (orgUnit != null) {
+        return getSelectedOrgUnits(user, program, orgUnit);
+      }
```

Comment:
Can we organize the if to make it clear that they are guard clauses even if we repeat some condition?

```suggestion
    if (orgUnitMode == null && orgUnit != null) {
        return getSelectedOrgUnits(user, program, orgUnit);
      }
      
      if (orgUnitMode == null && orgUnit == null) {
        return getAccessibleOrgUnits(user, program, orgUnit);
      }
      
```

Like this it is at a first sight what are the conditions

If the previous comment is implemented we do not even need these guard clauses